adafruit_st7735.py

```diff
@@ -66,16 +96,75 @@
     # fix on VTL
     b"\x3a\x01\x05" # COLMOD - 16bit color
-    b"\xe0\x10\x02\x1c\x07\x12\x37\x32\x29\x2d\x29\x25\x2B\x39\x00\x01\x03\x10" # _GMCTRP1 Gamma
-    b"\xe1\x10\x03\x1d\x07\x06\x2E\x2C\x29\x2D\x2E\x2E\x37\x3F\x00\x00\x02\x10" # _GMCTRN1
+)
+
+_INIT_R2_GREEN = (
     b"\x2a\x03\x02\x00\x81" # _CASET XSTART = 2, XEND = 129
     b"\x2b\x03\x02\x00\x81" # _RASET XSTART = 2, XEND = 129
+)
+
+_INIT_R2_RED = (
+    b"\x2a\x00\x00\x00\x7F" # _CASET XSTART = 0, XEND = 127
+    b"\x2b\x00\x00\x00\x9F" # _RASET XSTART = 0, XEND = 159
+)
+
+_INIT_R2_GREEN_144 = (
+    b"\x2a\x00\x00\x00\x7F" # _CASET XSTART = 0, XEND = 127
+    b"\x2b\x00\x00\x00\x7F" # _RASET XSTART = 0, XEND = 127
+)
+
+_INIT_R2_GREEN_160X80 = (
+    b"\x2a\x00\x00\x00\x4F" # _CASET XSTART = 0, XEND = 79
+    b"\x2b\x00\x00\x00\x9F" # _RASET XSTART = 0, XEND = 159
+)
+
+_INIT_R3 = (
+    b"\xe0\x10\x02\x1c\x07\x12\x37\x32\x29\x2d\x29\x25\x2B\x39\x00\x01\x03\x10" # _GMCTRP1 Gamma
+    b"\xe1\x10\x03\x1d\x07\x06\x2E\x2C\x29\x2D\x2E\x2E\x37\x3F\x00\x00\x02\x10" # _GMCTRN1
     b"\x13\x80\x0a" # _NORON
     b"\x29\x80\x64" # _DISPON
 )
 
+_INIT_R4_GREEN_160X80 = (
+    b"\x36\x01\x60" # _MADCTL Rotate to Landscape Mode
+)
+
+_INIT_R4_BLACK = (
+    b"\x36\x01\xC0" # _MADCTL Rotate to Landscape Mode
+)
+
+# pylint: disable=too-few-public-methods
 class ST7735(displayio.Display):
     """ST7735 driver for ST7735R Green tabs"""
-    # TODO(tannewt): Add support for Red tabs and non-R chips. https://github.com/adafruit/Adafruit-ST7735-Library/blob/master/Adafruit_ST7735.cpp
     def __init__(self, bus, *, width, height):
-        super().__init__(bus, _INIT_SEQUENCE, width=width, height=height, colstart=2)
+        init_sequence = _INIT_R1 + _INIT_R2_GREEN + _INIT_R3
+        super().__init__(bus, init_sequence, width=width, height=height, colstart=2, rowstart=1)
+
+class ST7735R_GREEN144(displayio.Display):
+    """ST7735 driver for ST7735R Green tabs 1.44-inch"""
+    def __init__(self, bus, *, width, height):
+        init_sequence = _INIT_R1 + _INIT_R2_GREEN + _INIT_R3
+        super().__init__(bus, init_sequence, width=width, height=height, colstart=2, rowstart=3)
+
+class ST7735R_RED(displayio.Display):
+    """ST7735 driver for ST7735R Red tabs"""
+    def __init__(self, bus, *, width, height):
+        init_sequence = _INIT_R1 + _INIT_R2_RED + _INIT_R3
+        super().__init__(bus, init_sequence, width=width, height=height)
+
+class ST7735R_BLACK(displayio.Display):
+    """ST7735 driver for ST7735R Black tabs"""
+    def __init__(self, bus, *, width, height):
+        init_sequence = _INIT_R1 + _INIT_R2_RED + _INIT_R3 + _INIT_R4_BLACK
+        super().__init__(bus, init_sequence, width=width, height=height)
+
+class ST7735B(displayio.Display):
+    """ST7735 driver for ST7735B"""
+    def __init__(self, bus, *, width, height):
+        super().__init__(bus, _INIT_SEQUENCE_B, width=width, height=height)
+
+class MINI160X80(displayio.Display):
```

Comment:
I'd like to see a separate library for Adafruit display products because it's the board and LCD pair that defines displayed resolution. The ST7735 functionality is separate of that. Once that is removed only ST7735B and ST7735R will remain. Then, in the adafruit display repo we can be very explicit about the setup on a per product number ID rather than tab color.

Reply:
I think I understand what you mean. Does that mean the rotation and color mode can still be passed after initialization?